File: radiation/grid_map_2d.py
"""
Copyright (c) 2015, The Regents of the University of California (Regents).
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

   3. Neither the name of the copyright holder nor the names of its
      contributors may be used to endorse or promote products derived
      from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

Please contact the author(s) of this library if you have any questions.
Authors: David Fridovich-Keil   ( dfk@eecs.berkeley.edu )
"""

###########################################################################
#
# Defines a grid map in 2D, with the specified dimensions.
# There is no true length scale in this structure; angles however are
# assumed to be metricly accurate.
#
# Implicitly, the values in the grid define a 2D non-homogeneous Poisson
# Process, which allows the random generation of possible sources. This
# is used to simulate measurements from arbitrary locations based on
# the current belief state.
#
###########################################################################

from source_2d import Source2D

import numpy as np
from numpy import linalg as LA
import math

import copy
from sets import ImmutableSet
from scipy.optimize import least_squares

class GridMap2D:
    def __init__(self, nrows, ncols, k, alpha=1.0):
        """
        Constructor. Takes in dimensions and number of sources k
        and generates a uniform prior. Alpha is a regularization
        parameter used in least squares optimization.
        """
        self.belief_ = np.ones((nrows, ncols), dtype=np.float) * k / (nrows*ncols)
        self.k_ = k
        self.alpha_ = alpha

        # Also, keep track of which voxels have been viewed at each point
        # in time, and the associated measurement.
        self.viewed_lists_ = []
        self.viewed_sets_ = []
        self.measurements_ = []

    def GenerateSources(self):
        """ Generates 'k' sources from the 'belief' prior. """
        prior = np.copy(self.belief_) / self.belief_.sum()

        # Choose a 'k' random numbers uniformly in [0, 1]. These will be treated
        # as evaluations of the CDF, and since they are uniform in [0, 1], the
        # points at which they occur are distributed according to 'prior'.
        cdf_evals = sorted(np.random.uniform(0.0, 1.0, self.k_))

        # Walk the 'prior' until we get to each 'cdf_eval' and record which voxel
        # we are in.
        sources = []
        current_cdf_index = 0

        cdf = 0.0
        for ii in range(prior.shape[0]):
            for jj in range(prior.shape[1]):
                cdf += prior[ii, jj]

                # Check if we just passed the next 'cdf_eval'.
                if cdf >= cdf_evals[current_cdf_index]:
                    # Generate a new source here.
                    source = Source2D(float(ii) + 0.5, float(jj) + 0.5)
                    sources.append(source)

                    # Return if we've got enough sources.
                    if len(sources) == self.k_:
                        return sources

                    # Increment 'current_cdf_index'.
                    current_cdf_index += 1

        # Should never get here.
        assert(False)
        return []

    def Update(self, sensor, solve=True):
        """
        Update belief about the world, given a sensor (with associated
        paramters, including position and orientation).

        For all voxels in range, create a rate update which is uniform
        and sums to the measurement value, then compute a weighted average
        at each point, trusting lower values more than higher ones.
        """
        measurement = sensor.Sense()
        if measurement > self.k_:
            print "Measured too many sources. Did not update."
            return False

        # Identify voxels that are in range. Store row-major flattened indices.
        in_view_list = []
        for ii in range(self.belief_.shape[0]):
            for jj in range(self.belief_.shape[1]):
                if sensor.VoxelInView(ii, jj):
                    in_view_list.append(ii*self.belief_.shape[1] + jj)

        # Set up constrained least squares problem and solve.
        self.viewed_lists_.append(in_view_list)
        self.viewed_sets_.append(ImmutableSet(in_view_list))
        self.measurements_.append(measurement)

        if solve:
            self.SolveLeastSquares()

        return True

    def SolveLeastSquares(self):
        """
        Solve the following constrained linear least squares problem:
                 min 0.5 * sum_i ((sum_{j in S_i} p_j) - z_i)^2
                     + 0.5 * alpha * ((sum_j p_j) - k)^2
                 s.t. all probabilities are between 0 and 1
        where S_i = {j : voxel v_j is in view at time i}.

        This is a linear objective, but the constraints make it slightly
        annoying to solve in closed form, so for now we solve with non-linear
        least squares optimization, providing derivatives wherever possible.
        """
        belief = self.belief_.flatten()

        # Pre-compute the (constant) Jacobian matrix.
        J = np.empty((len(self.viewed_sets_) + 1, len(belief)), dtype=np.float)
        J[:-1, :] = map(lambda ii : map(lambda jj : (jj in self.viewed_sets_[ii]),
                                        range(len(belief))),
                        range(len(self.viewed_sets_)))
        J[-1:, :] = math.sqrt(self.alpha_)

        # Compute the vector of residuals.
        def residuals(p, viewed_lists, measurements, k, alpha, jac):
            resids = [(p[S].sum() - z) for (S, z) in zip(viewed_lists, measurements)]
            resids.append(math.sqrt(alpha) * (p.sum() - k))
            return np.array(resids)

        # Compute the Jacobian matrix.
        def jacobian(p, viewed_lists, measurements, k, alpha, jac):
            return jac

        try:
            result = least_squares(residuals, belief, jac=jacobian, bounds=(0.0, 1.0),
                                   args=(self.viewed_lists_, self.measurements_,
                                         self.k_, self.alpha_, J),
                                   xtol=0.001,
                                   verbose=0)
            self.belief_ = np.reshape(result.x, self.belief_.shape)

        except Exception, e:
            pass


    def SimulateTrajectory(self, sensor, trajectory, niters=1):
        """
        Return expected map entropy after taking scans at each GridPose2D
        in the trajectory.

        Expectation is based on Monte Carlo simulation using the specified
        number of iterations.
        """
        entropy_total = 0

        # Save the current state. This could be speeded up.
        current_belief = np.copy(self.belief_)
        current_viewed_lists = list(self.viewed_lists_)
        current_viewed_sets = list(self.viewed_sets_)
        current_measurements = list(self.measurements_)

        # Monte Carlo simulation.
        for ii in range(niters):
            for pose in trajectory:
                # Set sensor pose.
                sensor.ResetPose(pose)

                # Simulate sensor measurement.
                assert self.Update(sensor, False)

            # Once all poses have been simulated, compute entropy.
            self.SolveLeastSquares()
            entropy_total += self.Entropy()

            # Restore state.
            self.belief_ = np.copy(current_belief)
            self.viewed_lists_ = list(current_viewed_lists)
            self.viewed_sets_ = list(current_viewed_sets)
            self.measurements_ = list(current_measurements)

        # Divide out by number of iterations.
        return entropy_total / float(niters)

    def SimulatePose(self, sensor, niters=1):
        """
        Return expected map entropy after receiving a measurement from
        the specified location/orientation. Expectation is based on
        Monte Carlo simulation using the specified number of iterations.
        """
        entropy_total = 0

        # Save the current belief state.
        current_belief = np.copy(self.belief_)

        # Iterate the specified number of times, accumulating total entropy.
        for ii in range(niters):
            assert self.Update(sensor)
            entropy_total += self.Entropy()

            # Reset belief state.
            self.belief_ = np.copy(current_belief)

        # Divide out by number of iterations.
        return entropy_total / float(niters)

    def Entropy(self):
        """
        Compute the entropy of the map. Since we model each voxel as a
        Bernoulli variable, independent of all others, we compute joint entropy
        as the sum of binary entropies at each voxels.
        """
        total_entropy = 0
        for ii in range(self.belief_.shape[0]):
            for jj in range(self.belief_.shape[1]):
                total_entropy += self.BernoulliEntropy(self.belief_[ii, jj])

        return total_entropy

    def BernoulliEntropy(self, p):
        """
        Binary entropy function (in nats) of a Bernoulli variable with
        parameter p.
        """
        assert (p >= 0.0) and (p <= 1.0)

        # For numerical stability, return 0 if p is nearly 0 or 1.
        if (p < 1e-8) or (p > 1.0 - 1e-8):
            return 0.0

        # Otherwise, use the definition of entropy.
        return -p * math.log(p) - (1.0 - p) * math.log(1.0 - p)

    def PoissonEntropy(self, rate):
        """
        Approximate entropy (in nats) of a Poisson variable with the
        given rate.
        """
        if rate < 1e-8:
            return 0.0

        sum_max = min(6, 2*int(round(rate)))
        entropy = rate * (1.0 - math.log(rate))

        # Add on sum_max extra terms.
        extra_terms = 0
        rate_ii = rate
        fact_ii = 1
        for ii in range(2, sum_max + 1):
            rate_ii *= rate
            fact_ii *= float(ii)
            extra_terms += rate_ii * math.log(fact_ii) / fact_ii

        # Scale extra terms and add to entropy.
        entropy += math.exp(-rate) * extra_terms
        return entropy

File: radiation/cpp/src/grid_map_2d.cpp
/*
 * Copyright (c) 2015, The Regents of the University of California (Regents).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    1. Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 *    3. Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Please contact the author(s) of this library if you have any questions.
 * Author: David Fridovich-Keil   ( dfk@eecs.berkeley.edu )
 */

///////////////////////////////////////////////////////////////////////////////
//
// Defines a 2D grid map.
//
///////////////////////////////////////////////////////////////////////////////

#include <grid_map_2d.h>
#include <encoding.h>
#include <cost_functors.h>

#include <ceres/ceres.h>
#include <glog/logging.h>
#include <math.h>
#include <algorithm>
#include <unordered_map>

namespace radiation {

  GridMap2D::~GridMap2D() {}
  GridMap2D::GridMap2D(unsigned int num_rows, unsigned int num_cols,
                       unsigned int num_sources, double regularizer)
    : num_rows_(num_rows), num_cols_(num_cols),
      num_sources_(num_sources), regularizer_(regularizer),
      rng_(rd_()) {

    // Initialize belief matrix to be uniform.
    belief_ = Eigen::MatrixXd::Ones(num_rows_, num_cols_) * num_sources_;
    belief_ /= num_rows_ * num_cols_;
  }

  // Generate random sources according to the current belief state.
  bool GridMap2D::GenerateSources(std::vector<Source2D>& sources) {
    const double total_belief = belief_.sum();
    std::uniform_real_distribution<double> unif(0.0, 1.0);

    // Choose 'num_sources_' random numbers in [0, 1], which will be sorted
    // and treated as evaluations of the CDF. Since they are uniform, the points
    // at which they occur ar distributed according to the current belief.
    std::vector<double> cdf_evals;
    for (size_t ii = 0; ii < num_sources_; ii++)
      cdf_evals.push_back(unif(rng_));

    std::sort(cdf_evals.begin(), cdf_evals.end());

    // Walk the current belief distribution until we get to each 'cdf_eval'
    // and record which voxel we are in.
    sources.clear();
    unsigned int current_index = 0;
    double current_cdf = 0.0;

    for (unsigned int ii = 0; ii < num_rows_; ii++) {
      for (unsigned int jj = 0; jj < num_cols_; jj++) {
        current_cdf += belief_(ii, jj) / total_belief;

        // Check if we just passed the next 'cdf_eval'.
        while (current_cdf >= cdf_evals[current_index]) {
          // Generate a new source here.
          sources.push_back(Source2D(ii, jj));

          // Return if we have enough sources.
          if (sources.size() == num_sources_)
            return true;

          // Increment 'current_index'.
          current_index++;
        }
      }
    }

    // Should never get here.
    return false;
  }

  // Generate conditional distribution [P_{Z|X}] and entropy vector [h_{M|Z}],
  // where the (i, j)-entry of [P_{Z|X}] is the normalized frequency of
  // observing measurement i given trajectory j, and the i-entry of [h_{M|Z}]
  // is the entropy of M given measurement i, starting from the given pose.
  void GridMap2D::GenerateConditionals(
     unsigned int num_samples, unsigned int num_steps, const GridPose2D& pose,
     double sensor_fov, Eigen::MatrixXd& pzx, Eigen::VectorXd& hmz,
     std::vector<unsigned int>& trajectory_ids) const {
    // Compute the number of possible maps and measurements.
    const unsigned int kNumMeasurements = pow(num_sources_ + 1, num_steps);
    const unsigned int kNumMaps = pow(num_rows_ * num_cols_, num_steps);

    // Create a hash table to keep track of counts for each trajectory.
    std::unordered_map<unsigned int, Eigen::VectorXd> zx_samples;

    // Create an empty matrix to store the joint distribution [P_{M, Z}].
    Eigen::MatrixXd zm_joint =
      Eigen::MatrixXd::Zeros(kNumMeasurements, kNumMaps);

    // Generate a ton of sampled data.
    for (unsigned int ii = 0; ii < num_samples; ii++) {
      // Generate random sources on the grid according to the current 'belief',
      // and compute a corresponding 'map_id' number based on which grid cells
      // the sources lie in.
      std::vector<Source2D> sources;
      if (!GenerateSources(sources)) {
        VLOG(1) << "Unable to generate sources. Skipping this sample.";
        continue;
      }

      const unsigned int map_id = EncodeMap(sources, num_rows_, num_cols_);

      // Pick a random trajectory starting at the given pose. At each step,
      // take a measurement and record the data.
      GridPose2D current_pose = pose;
      std::vector<Movement2D> movements;
      std::vector<unsigned int> measurements;
      while (movements.size() < num_steps) {
        const Movement2D step;
        if (current_pose.MoveBy(step)) {
          movements.push_back(step);

          const Sensor2D sensor(current_pose, sensor_fov);
          measurements.push_back(sensor.Sense(sources));
        }
      }

      // Compute trajectory and measurement sequence ids.
      const unsigned int trajectory_id = EncodeTrajectory(movements);
      const unsigned int measurement_id =
        EncodeMeasurements(measurements, num_sources_);

      // Record this sample in the 'zm_joint' matrix.
      zm_joint(measurement_id, map_id) += 1.0;

      // Record this sample in the 'zx_samples' hashmap.
      if (zx_samples.count(trajectory_id) > 0)
        zx_samples.insert({trajectory_id,
                           Eigen::VectorXd::Zeros(kNumMeasurements)});

      zx_samples[trajectory_id](measurement_id) += 1.0;
    }

    // Convert 'zx_samples' into a matrix joint distribution.
    Eigen::MatrixXd zx_joint =
      Eigen::MatrixXd::Zeros(kNumMeasurements, zx_samples.size());
    std::vector<unsigned int> trajectory_ids;
    trajectory_ids.reserve(zx_samples.size());

    unsigned int idx = 0;
    for (auto pair = zx_samples.begin(); pair != zx_samples.end(); pair++) {
      trajectory_ids.push_back(pair->first);
      zx_joint.col(idx++) = pair->second;
    }

    // Normalize so that all rows sum to unity in both distributions.
    for (unsigned int ii = 0; ii < kNumMeasurements; ii++) {
      const double zm_row_sum = zm_joint.row(ii).sum();
      if (zm_row_sum < 1.0)
        VLOG(1) << "Encountered measurement with no support in P_{Z|M}.";
      else
        zm_joint.row(ii) /= zm_row_sum;

      const double zx_row_sum = zx_joint.row(ii).sum();
      if (zx_row_sum < 1.0)
        VLOG(1) << "Encountered measurement with no support in P_{Z|X}.";
      else
        zx_joint.row(ii) /= zx_row_sum;
    }

    // Compute [h_{M|Z}], the conditional entropy vector.

  }

  // Take a measurement from the given sensor and update belief accordingly.
  bool GridMap2D::Update(const Sensor2D& sensor,
                         const std::vector<Source2D>& sources,
                         bool solve) {
    const unsigned int measurement = sensor.Sense(sources);
    CHECK(measurement <= num_sources_);

    // Identify all voxels in range and store.
    std::vector<unsigned int> voxels;
    for (size_t ii = 0; ii < num_rows_; ii++) {
      for (size_t jj = 0; jj < num_cols_; jj++) {
        if (sensor.VoxelInView(ii, jj))
          voxels.push_back(ii + jj * num_rows_);
      }
    }

    viewed_.push_back(voxels);
    measurements_.push_back(measurement);

    // Maybe solve.
    if (solve)
      return SolveLeastSquares();

    return true;
  }

  // Compute entropy.
  double GridMap2D::Entropy() const {
    double entropy = 0.0;

    for (size_t ii = 0; ii < num_rows_; ii++) {
      for (size_t jj = 0; jj < num_cols_; jj++) {
        const double p = belief_(ii, jj);
        const double bernoulli_entropy = (p > 1e-8 && p < 1.0 - 1e-8) ?
          (-p * log(p) - (1.0 - p) * log(1.0 - p)) : 0.0;

        entropy += bernoulli_entropy;
      }
    }

    return entropy;
  }

  // Solve least squares problem to update belief state.
  bool GridMap2D::SolveLeastSquares() {
    // Create a non-linear least squares problem.
    ceres::Problem problem;

    // Add residual blocks for each set of observed voxels and their
    // associated measurements. Note that 'belief_' is laid out in column-
    // major order by default, so we can just use its underlying structure
    // as the optimization variable.
    for (size_t ii = 0; ii < viewed_.size(); ii++) {
      problem.AddResidualBlock(
        BeliefError::Create(num_rows_, num_cols_,
                            &viewed_[ii], measurements_[ii]),
        NULL, /* squared loss */
        belief_.data());
    }

    // Add a final residual block to enforce consistancy across the entire grid,
    // i.e. that the expected number of sources matches the specified number.
    problem.AddResidualBlock(
      BeliefRegularization::Create(num_rows_, num_cols_, num_sources_,
                                   regularizer_ * viewed_.size()),
      NULL, /* squared loss */
      belief_.data());

    // Set bounds constraints. Each voxel's belief should be a probability
    // between 0 and 1.
    for (int ii = 0; ii < num_rows_ * num_cols_; ii++) {
      problem.SetParameterLowerBound(belief_.data(), ii, 0.0);
      problem.SetParameterUpperBound(belief_.data(), ii, 1.0);
    }

    // Set up solver options.
    ceres::Solver::Summary summary;
    ceres::Solver::Options options;
    options.linear_solver_type = ceres::DENSE_QR;
    options.function_tolerance = 1e-16;
    options.gradient_tolerance = 1e-16;
    options.trust_region_strategy_type = ceres::LEVENBERG_MARQUARDT;

    // Solve and return.
    ceres::Solve(options, &problem, &summary);

    return summary.IsSolutionUsable();
  }

  // Get a reference to immutable 'belief'.
  const Eigen::MatrixXd& GridMap2D::GetImmutableBelief() const {
    return belief_;
  }

} // namespace radiation
